Update point source analysis endpoint to support PA source option

rachelekm opened the pull request:
## Overview

Adds a new endpoint to request a point source analysis with a given datasource override, enabling users to specify using PA point source data for an analysis. See notes on decision to separate these into two separate endpoints vs one endpoint with an optional path parameter. Preserves default convention of using CONUS datasource if outside the Delaware river basin and the DRB if within it if no datasource override is provided.

Connects #3693 

### Demo

Failing analysis job when provided datasource is outside the given area of interest:
<img width="1344" height="700" alt="Screenshot 2025-09-08 at 3 45 05 PM" src="https://github.com/user-attachments/assets/d6e8d696-b3d9-4fd4-aec2-720847d62a3a" />


Failing analysis job when provided datasource is not "conus", "pa", or "drb":
<img width="1235" height="690" alt="Screenshot 2025-09-08 at 3 45 21 PM" src="https://github.com/user-attachments/assets/8e79db37-3d22-4019-9364-0b9eaca674b5" />


Successful point source analysis for PA, including updated display name:
<img width="1294" height="832" alt="Screenshot 2025-09-08 at 5 47 01 PM" src="https://github.com/user-attachments/assets/c720673e-7557-4d9a-aa22-97a4f3ab55ce" />


### Notes

Ideally the point source analysis would remain one endpoint with an optional path parameter, "datasource". However, this caused a breaking error with Swagger/OpenAPI as [any parameter in the path is required](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#fixed-fields-7) in their spec. If I remove the swagger parameter schema for this endpoint the path displays as `/{var}` in the docs and there's lacking descriptions for the param input. I tried including `datasource_override` in the request body instead, but since this is a shared schema across multiple different analyses (in the frontend and backend) it was a lot of lift to manage side effects and seemed prone to regressions. Separating the endpoints felt like a better separation of concerns with appropriate documentation and easier validation, albeit some redundancy. I'm curious what you think!

## Testing Instructions

 * `vagrant up`
 * Go to http://localhost:8000/api/docs  and login
 * Confirm new `/analyze/pointsource/{datasource}/` endpoint
 * Test endpoint with area of interest outside of DRB, below example around Pittsburgh:
 ```
 {
    "area_of_interest": {
        "type": "MultiPolygon",
        "coordinates": [
            [
                [
                    [
                        -75.17532348632812,
                        40.04312188564563
                    ],
                    [
                        -75.26390075683594,
                        40.07491742068019
                    ],
                    [
                        -75.38028717041016,
                        39.96475227412719
                    ],
                    [
                        -75.2508544921875,
                        39.84439329944007
                    ],
                    [
                        -75.17051696777344,
                        39.94080268676023
                    ],
                    [
                        -75.17532348632812,
                        40.04312188564563
                    ]
                ]
            ]
        ]
    }
}
 ```
* Add "drb" as datasource parameter and execute
* Confirm error response as area or interest outside of DRB
* Add "pa" as datasource parameter and execute
* Confirm request is successful and copy job id
* Past id in job_uuid input of `/jobs/{job_uuid}/` endpoint and execute
 * Confirm job successful and, once complete, response include point sources from the `ms_pointsource_pa` table
 * Confirm display name and name include the substring "(PA)"
 * Run again, this time with `conus` as the overriding datasource
 * Confirm successful and response includes point sources from the `ms_pointsource` table
 *  Confirm display name and name include the substring "(CONUS)"
 * Use same data object to request a point source analysis with no override, `/analyze/pointsource/`
 * Confirm no regressions and categories response from completed job matches that last `conus` job with same area of interest
 * Run `/analyze/pointsource/` with an area of interest in the drb
 * Confirm no regressions and, once job complete, confirm point sources are from the `ms_pointsource_drb` table

rachelekm committed:
Enable overriding datasource for pointsource analysis

Updates analysis calculation to enable an optional overriding source to retrieve data. Preserves logic to determine table source by AOI, but overrides with respective table if datasource exists.

rachelekm committed:
Add endpoint to support analysis with override datasource

rajadain reviewed (changes requested):
This is a good start. I agree with the two endpoint approach.

Let's rename the old one to be `start_analyze_pointsource_deprecated` and the new one to be `start_analyze_pointsource`.

Let's also include a deprecation message in the old point source endpoint like this:

https://github.com/WikiWatershed/model-my-watershed/blob/a7caeabcfca5cd665c6cbb78df4db67d1a977da2/src/mmw/apps/geoprocessing_api/views.py#L2247-L2251

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1053,7 +1053,8 @@ def start_analyze_animals(request, format=None):
                      request_body=schemas.ANALYZE_REQUEST,
-                     responses={200: schemas.JOB_STARTED_RESPONSE})
+                     responses={200: schemas.JOB_STARTED_RESPONSE},
+                     operation_id='analyze_pointsource')
```
Ah interesting. I would recommend splitting this change into its own commit, and adding this context to the commit message, so it's easier to understand and remember.

rajadain replied:
Ah I see this now also described in the **Notes** section. I think we have enough explanation, so I'll leave the separation into its own commit to your discretion.

rachelekm replied:
I like the idea of pulling it out into its own commit, that is now in 37f796966287346839abbf3beaaed6314f2153e6

rachelekm committed:
Set overriding endpoint id for Swagger UI

Swagger/OpenAPI ignores path params when creating internal endpoint ids, so both of point source endpoints are assigned the same identifier, analyze_pointsource_create, by default. This causes UI issues in the Swagger docs. This commit explicitly sets the identifier using operation_id for the deprecated endpoint to make both ids unique and clarify endpoint role in docs.

rachelekm commented on src/mmw/mmw/settings/layer_settings.py:
```diff
@@ -724,6 +724,19 @@ def _make_lulc_config(year):
+    'CONUS': 'ms_pointsource',
+    'DRB_SIMPLE': 'ms_pointsource_drb',
+    'PA_SIMPLE': 'ms_pointsource_pa'
+}
```
Updated in 164c6201cf35566108da5af429cb113c15679fe2

rachelekm commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -187,25 +187,35 @@ def calculate_avg_slope(slope, length):
+            or datasource == settings.POINT_SOURCE_OPTIONS["pa"]
+        )
+        table_override = settings.POINT_SOURCE_OPTIONS[datasource]
+    table_name = get_point_source_table(drb_or_pa, table_override)
```
Updated in 164c6201cf35566108da5af429cb113c15679fe2

rachelekm commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -187,25 +187,35 @@ def calculate_avg_slope(slope, length):
-    from the `ms_pointsource` or `ms_pointsource_drb` table to display
-    in the Analyze tab.
+    from the `ms_pointsource`, `ms_pointsource_drb` or
+    `ms_pointsource_drb` table to display in the Analyze tab.
```
Fixed in 164c6201cf35566108da5af429cb113c15679fe2

rachelekm commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -187,25 +187,35 @@ def calculate_avg_slope(slope, length):
           WHERE ST_Intersects(geom, ST_SetSRID(ST_GeomFromText(%s), 4326))
           '''
-
+    print('HEY', drb_or_pa, table_name)
```
Fixed in 58722cc65db6a5a2ba67b9716d41939ab6fc9622

rachelekm commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1111,6 +1112,75 @@ def start_analyze_pointsource(request, format=None):
+    user = request.user if request.user.is_authenticated else None
+    if datasource not in settings.POINT_SOURCE_OPTIONS:
+        return Response(
+            {"detail": 'Invalid datasource requested.'},
```
Fixed in 58722cc65db6a5a2ba67b9716d41939ab6fc9622

rachelekm commented:
Thanks for the review, @rajadain! I've addressed those comments and also updated the endpoints so that the current one is now `start_analyze_pointsource_deprecated` and the new one is `start_analyze_pointsource`. Calling the deprecated endpoint also now includes a deprecation message:
<img width="1138" height="284" alt="Screenshot 2025-09-10 at 2 02 27 PM" src="https://github.com/user-attachments/assets/24a4efba-77d3-4060-b524-f60fec31c276" />

This is ready for another look!

rajadain commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -226,8 +231,10 @@ def point_source_pollution(geojson):
-        'name': 'pointsource',
+        'displayName': 'Point Source ({})'.format(datasource.upper())
+        if datasource else 'Point Source',
+        'name': 'pointsource ({})'.format(datasource.upper())
```
No spaces allowed here.

```suggestion
        'name': f'pointsource_{datasource}'
```

rajadain commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -226,8 +231,10 @@ def point_source_pollution(geojson):
     return {
-        'displayName': 'Point Source',
-        'name': 'pointsource',
+        'displayName': 'Point Source ({})'.format(datasource.upper())
```
f-strings are now available for this project

```suggestion
        'displayName': f'Point Source ({datasource.upper()})'
```

rajadain commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -187,25 +187,30 @@ def calculate_avg_slope(slope, length):
-    sql = f'''
+    table_name = get_point_source_table(geom.within(DRB), datasource)
+    # facilityname col exists for PA and DRB tables
+    has_facilityname = table_name.endswith("_pa") or table_name.endswith(
+        "_drb"
+    )
+    sql = f"""
```
The project convention is to use single quotes unless otherwise necessary

```suggestion
    has_facilityname = table_name.endswith('_pa') or table_name.endswith(
        '_drb'
    )
    sql = f'''
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
     area_of_interest, wkaoi, msg = _parse_analyze_input(request,
                                                         perimeter='CONUS')
 
+    deprecation_message = [
```
Let's append this to `msg` which may have other messages:


```suggestion
    msg.append(
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
+        'endpoint with a valid datasource instead.'
+        'Valid datasource options must be one of: {}'.format(
+            ', '.join(settings.POINT_SOURCES.keys())),
+    ]
```
```suggestion
    )
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
+
     return start_celery_job([
         tasks.analyze_pointsource.s(area_of_interest)
+    ], area_of_interest, user, None, deprecation_message)
```
```suggestion
    ], area_of_interest, user, messages=msg)
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
+        {
+            "survey": {
+                "displayName": "Point Source {datasource option}",
+                "name": "pointsource {datasource option}",
```
```suggestion
                "name": "pointsource_{datasource option}",
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
+    user = request.user if request.user.is_authenticated else None
+    if datasource not in settings.POINT_SOURCES:
+        raise ValidationError(
+            f"Invalid pointsource datasource: {datasource}."
```
```suggestion
            f'Invalid pointsource datasource: {datasource}.'
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
+                '", "'.join(settings.POINT_SOURCES.keys())
+            )
+        )
+    perimeter = settings.POINT_SOURCES[datasource]["perimeter"]
```
```suggestion
    perimeter = settings.POINT_SOURCES[datasource]['perimeter']
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
+
+    return start_celery_job([
+        tasks.analyze_pointsource.s(area_of_interest, datasource)
     ], area_of_interest, user)
```
```suggestion
    ], area_of_interest, user, messages=msg)
```

rajadain commented on src/mmw/apps/geoprocessing_api/views.py:
```diff
@@ -1106,8 +1107,86 @@ def start_analyze_pointsource(request, format=None):
+    deprecation_message = [
+        'Using this endpoint without a datasource path parameter will be '
+        'deprecated in an upcoming release. Please switch to using this '
+        'endpoint with a valid datasource instead.'
```
```suggestion
        'endpoint with a valid datasource instead. '
```

rajadain reviewed (commented):
This looks much better! Nice job on the refactor! Unfortunately we don't have linting for this project, so leaving a number of little suggestions. I'm testing the outputs now, will report back.

rajadain commented:
I tested with these three HUC-10s:

- Chicago `0712000407`
    <img width="1854" height="1048" alt="image" src="https://github.com/user-attachments/assets/d10fa218-2c68-4d6b-8aeb-98f452c9e0c4" />
- Pittsburgh `0501000903`
    <img width="1854" height="1048" alt="image" src="https://github.com/user-attachments/assets/53d35c29-5eb8-45a1-8e1a-3aa9465e3164" />
- Philadelphia `0204020204`
    <img width="1854" height="1048" alt="image" src="https://github.com/user-attachments/assets/77e202fa-576f-4f90-abf9-eb9bc58a52ed" />

I confirmed that the numbers for Chicago were identical between deprecated and conus, for Pittsburgh between deprecated and conus, and for Philadelphia between deprecated and drb.

I further confirmed that Chicago got a 400 for trying pa and drb, and Philadelphia got a 400 for trying pa (since much of that area of interest is in New Jersey). Given that last bit, we may potentially have to up the boundary buffer, but we will leave that for later if Barry asks.

The functionality for this PR is correct! Nice work. Will take a final look at the code once all the suggestions have been resolved.

rajadain commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -187,25 +187,30 @@ def calculate_avg_slope(slope, length):
           WHERE ST_Intersects(geom, ST_SetSRID(ST_GeomFromText(%s), 4326))
-          '''
-
+          """
```
I missed a spot

```suggestion
          '''
```

rachelekm replied:
This one should be captured in 3790ff5719c597ec837c674cc8a0b92ec0d7cd25

rachelekm commented:
Thanks, @rajadain! Thank you for those suggestions and great that its working with the three HUC-10s. I implemented your suggestions (I grabbed that last one too) and they should all be resolved in the last 3 fixups I just pushed up. This is ready for a final look!

rajadain reviewed (approved):
Added one last fixup to get CI to work. Everything looks good! Nice work on all the iterations!

Friendly reminder to squash the fixups before merging.

rachelekm commented:
Thank you! Will rebase and squash and then merge on successful CI 👍
